x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlTasks.java

```diff
@@ -41,6 +42,17 @@
         new PersistentTasksCustomMetadata.Assignment(null,
             "persistent task will not be assigned as a feature reset is in progress.");
 
+    // When a master node action is executed and there is no master node the transport will wait
+    // for a new master node to be elected and retry against that, but will only wait as long as
+    // the configured master node timeout. In ESS rolling upgrades can be very slow, and a cluster
+    // might not have a master node for an extended period - over 1 minute was observed in the test
+    // that led to this constant being added. The default master node timeout is 30 seconds, which
+    // makes sense for user-invoked actions. But for actions invoked by persistent tasks that will
+    // cause the persistent task to fail if they time out waiting for a master node we need to wait
+    // longer, because failing the task just because a rolling upgrade was slow defeats the point
+    // of the task being "persistent".
+    public static final TimeValue PERSISTENT_TASK_MASTER_NODE_TIMEOUT = TimeValue.timeValueSeconds(900);
```

Comment:
I wonder why we chose 15 minutes here. What happens when it times out? Is there a risk that we end up accumulating multiple copies of the same master task in a queue somewhere, or otherwise consuming increasing resources over time while waiting? If not, would `TimeValue.MAX_VALUE` be appropriate instead?

In my experience if the cluster has no master for more than a minute or so then it is often hours or days before it recovers, typically because the remedy needs manual intervention. We do know it will eventually recover (or die trying) so do we lose anything by being more patient here?

If we do simply retry on a timeout then IMO it'd be better just to wait. On the other hand if we turn a timeout into a permanent failure then that would prevent automatic recovery when the cluster finally heals, so again IMO it'd be better just to wait.

Reply:
> If not, would `TimeValue.MAX_VALUE` be appropriate instead?

Yes, probably.  Effectively we want the ML persistent tasks to sit in limbo in their current state until the cluster recovers instead of failing.

The only reservation I have about `TimeValue.MAX_VALUE` is that it might expose some sort of overflow error somewhere in the code that uses the timeout.  For example if there's some code somewhere that's doing additions or subtractions using the timeout without considering that it might be right on the limit of what a `long` can store.

So how about 30 days as the timeout?  If somebody doesn't fix a cluster for a month then it's probably reasonable that they need to revisit their ML jobs.

> In my experience if the cluster has no master for more than a minute or so then it is often hours or days before it recovers

We do have logs from an ESS cluster that was upgraded where the cluster was without a master for somewhere between 30 and 66 seconds, and the only thing that went wrong was that an ML job failed.  So maybe a little over 30 seconds is common in ESS but nobody complains because they expect the upgrade to disrupt the cluster for a while and usually there are no permanent problems as a result.